mri/operators/fourier/utils/orc_coefficients.py

```diff
@@ -0,0 +1,278 @@
+##########################################################################
+
+"""
+Functions to compute different kinds of coefficients used in
+Off-Resonance Correction.
+"""
+
+
+import numpy as np
+import scipy.sparse.linalg as ssl
+import sklearn.cluster as sc
+
+
+def create_histogram(field_map, mask, weights="full", n_bins=1000):
+    """ Create the weighted histogram of the field map covered by the mask
+
+    Parameters
+    ----------
+    field_map: numpy.ndarray
+        B0 field inhomogeneity map (in Hz)
+    mask: numpy.ndarray
+        Mask describing the regions to consider during correction
+    weights: {'full', 'sqrt', 'log', 'ones'}
+        Weightning policy for the field map histogram (default 'full')
+    n_bins: int
+        Number of bins for the field map histogram (default 1000)
+
+    Returns
+    -------
+    histogram_centers: numpy.ndarray
+        Central values of each bin
+    histogram_counts: numpy.ndarray
+        Weights of elements in each bin according to args
+    """
+    # Compute the histogram centers and counts
+    field_map = field_map[np.where(mask)]
+    histogram_counts, histogram_edges = np.histogram(field_map, n_bins)
+    histogram_centers = (histogram_edges + (histogram_edges[1]
+                                            - histogram_edges[0]) / 2)[:-1]
+
+    # Change the weightning according to args
+    if (weights == "ones"):
+        histogram_counts = np.array(histogram_counts != 0).astype(int)
+    elif (weights == "sqrt"):
+        histogram_counts = np.sqrt(histogram_counts)
+    elif (weights == "log"):
+        histogram_counts = np.log(1 + histogram_counts)
+    elif (weights != "full"):
+        raise NotImplementedError("Unknown weightning: {}".format(weights))
+
+    return histogram_centers, histogram_counts
+
+
+def create_variable_density(centers, counts, L):
+    """ Find a reduced histogram with variable density from previous histogram
+
+    Parameters
+    ----------
+    centers: numpy.ndarray
+        Central value of each bin from previous histogram
+    counts: numpy.ndarray
+        Number of elements in each bin from previous histogram
+    L: int
+        Number of interpolators defining the size of the reduced histogram
+
+    Returns
+    -------
+    centers: numpy.ndarray
+        New bin centers of the reduced histogram
+    """
+
+    # Compute kmeans to get custom centers
+    km = sc.KMeans(n_clusters=L, random_state=0)
+    km = km.fit(samples.reshape((-1, 1)), sample_weight=counts)
+    centers = np.array(sorted(km.cluster_centers_)).flatten()
+
+    return centers
+
+
+def compute_mfi_coefficients(field_map, time_vec, mask, L=15,
+                             weights="full", n_bins=1000):
+    """
+    This function implements the correction weights described
+    in :cite: `man1997` and extended in :cite: `fessler2005`
+    called 'Multi-frequency interpolation' coefficients
+
+    Parameters
+    ----------
+    field_map: numpy.ndarray
```

Comment:
I dont think we need to repeat the documentation. Let us just have these as internal function and ask the user to refer to the ORC Wrapper for details. For now, I feel this just adds maintainence overhead. Its your call..
We can later think of how to repeat the documentation in sphinx generation pipelines. I am sure that it is doable.

Reply:
YOu can change function name to `def _compute_...` in case you make this internal function and you are sure that no user will use it.